lib/stdlib/src/erl_tar.erl

```diff
@@ -1253,37 +1259,38 @@ parse_numeric(<<First, _/binary>> = Bin) ->
 
 parse_octal(Bin) when is_binary(Bin) ->
     %% skip leading/trailing zero bytes and spaces
-    do_parse_octal(Bin, <<>>).
-do_parse_octal(<<>>, <<>>) ->
-    0;
+    do_parse_octal(Bin, 0).
+
 do_parse_octal(<<>>, Acc) ->
-    case io_lib:fread("~8u", binary:bin_to_list(Acc)) of
-        {error, _} -> throw({error, invalid_tar_checksum});
-        {ok, [Octal], []} -> Octal;
-        {ok, _, _} -> throw({error, invalid_tar_checksum})
-    end;
-do_parse_octal(<<$\s,Rest/binary>>, Acc) ->
+    Acc;
+do_parse_octal(<<$\s, Rest/binary>>, Acc) ->
     do_parse_octal(Rest, Acc);
 do_parse_octal(<<0, Rest/binary>>, Acc) ->
     do_parse_octal(Rest, Acc);
 do_parse_octal(<<C, Rest/binary>>, Acc) ->
-    do_parse_octal(Rest, <<Acc/binary, C>>).
+    Digit = C - $0,
+    if
+        0 =< Digit, Digit < 8 ->
+            do_parse_octal(Rest, Acc bsl 3 bor Digit);
+        true ->
+            throw({error, invalid_tar_checksum})
+    end.
 
 parse_string(Bin) when is_binary(Bin) ->
-    do_parse_string(Bin, <<>>).
-do_parse_string(<<>>, Acc) ->
-    case unicode:characters_to_list(Acc) of
+    N = strlen(Bin, 0),
+    <<Prefix:N/binary,_/binary>> = Bin,
+    case unicode:characters_to_list(Prefix) of
         Str when is_list(Str) ->
             Str;
         {incomplete, _Str, _Rest} ->
-            binary:bin_to_list(Acc);
+            binary_to_list(Bin);
```

Comment:
Does it make more sense here to use `Prefix` rather than `Bin` just so that it's clear we're returning bytes that are non-zero rather than anything that was passed in?

Functionally I don't think it'd make much of a difference so up to you

Reply:
I kept the original handling of incomplete Unicode characters. I am not sure that it the right thing to do, but I don't want to change any behavior of `erl_tar` in this pull request.